feat: use custom attribute types in schema

ffMathy opened the pull request:
Look at and merge #41 first, as this builds on top of that.

<!--
  [For internal use] 
  Copy the task id from the bottom of the sidebar and paste it after FTRACK-

  Resolves FTRACK-

-->

- [X] I have added automatic tests where applicable
- [X] The PR title is suitable as a release note
- [X] The PR contains a description of what has been changed
- [X] The description contains manual test instructions

## Changes
This change builds on top of #41 and makes the "custom_attributes" property strong-typed. Note that the custom attribute types are not exported, so they are just "private" in a sense for providing a better experience.

Here are some examples:

```typescript
const shot: Shot = ...;
const asset: Asset= ...;

//this is now a "number", because it recognizes the key.
shot.custom_attributes.find(x => x.key === "fps")?.value;

//the following throws an error, because 'aps' is not a valid custom attribute.
shot.custom_attributes.find(x => x.key === "aps")?.value;

//the following throws an error, because 'fps' is not a valid custom attribute for the type "asset".
asset.custom_attributes.find(x => x.key === "fps")?.value;
```

Combined with TypeScript 5.5 that just released, this also works very well with `filter` operations being narrowed correctly.

## Test
Automated tests.

ffMathy committed:
feat: add types to custom attributes

ffMathy committed:
Merge remote-tracking branch 'origin/main' into feature/add-types-to-custom-attributes

ffMathy committed:
feat: add strong-typed custom attributes

ffMathy commented:
@gismya is not sure if you saw this one yet or everybody is ok vacation, but I do believe it is valuable to get out there 😁

gismya commented:
I've seen it and I like it. Currently working on other things, so I've asked @jimmycallin to test it 🙂

ffMathy commented:
I guess people are also on vacation? 😁❤️

jimmycallin commented:
> I guess people are also on vacation? 😁❤️

Mainly busy with other things :) Hope to get to it this week!

ffMathy committed:
fix: re-add eslint to test step

ffMathy committed:
Merge branch 'feature/add-types-to-custom-attributes' into feature/use-custom-attribute-types-in-schema

gismya committed:
Merge branch 'main' into feature/use-custom-attribute-types-in-schema

gismya commented on source/emitCustomAttributes.ts:
```diff
@@ -0,0 +1,80 @@
+      export type RuntimeCustomAttributeConfigurationLabel = RuntimeCustomAttributeConfiguration["label"];
+  
+      ${customAttributes
+        .map(x => `type ${getCustomAttributeTypeNameFromAttribute(x)} = Omit<ContextCustomAttributeValue, 'key' | 'value'> & {
```
I'd prefer using a `BaseCustomAttributeValue` that doesn't have `key` and `value`, and extending from that. Preferably then that's what's used to create the untyped `ContextCustomAttributeValue` type, but if it's easier you can also create that separately as `Omit<ContextCustomAttributeValue, 'key' | 'value'>` and only use it for the typed custom attributes.

ffMathy replied:
This is now resolved.

gismya commented on source/__snapshots__/default-highway-test.snap:
```diff
@@ -1552,6 +1644,98 @@ export type RuntimeCustomAttributeConfigurationName =
 export type RuntimeCustomAttributeConfigurationLabel =
   RuntimeCustomAttributeConfiguration["label"];
 
+type CustomAttributeValue_somedynamicenumerator = Omit<
```
Would it be possible to use a generic instead of suffixes? I haven't tried, but something like:

```
interface TypedCustomAttributeValueMap {
  somedynamicenumerator: string[];
  someboolean: boolean;
}

type TypedCustomAttributeValue<K extends keyof TypedCustomAttributeValueMap> = Omit<
  ContextCustomAttributeValue,
  "key" | "value"
> & {
  key: K;
  value: ValueMap[K];
};

```

gismya replied:
That also removes the need for the other comment - having a single ```
Omit<
  ContextCustomAttributeValue,
  "key" | "value"
>
``` feels fine

gismya commented on source/__snapshots__/default-highway-test.snap:
```diff
@@ -1120,16 +1120,108 @@ export interface Priority {
-export type Shot = TypedContext<"Shot">;
-export type Sequence = TypedContext<"Sequence">;
-export type Information = TypedContext<"Information">;
+export type Milestone = Omit<TypedContext<"Milestone">, "custom_attributes"> & {
```
More of a long shot, but it could be possible to use a map here as well to add the custom attributes straight in `TypedContext` and not have to create new types.

gismya replied:
It would get rid of having to create new types and have both `TypedContext<"Milestone">` and `Milestone` with only subtle difference

gismya replied:
Don't have time to develop it fully, but something like this is what I'm imagining:
```
export interface TypedContextSubtypeMap {
  TypedContext: TypedContext;
  Milestone: Milestone;
}

export interface TypedContextCustomAttributesMap {
  TypedContext: Array<
    TypedCustomAttributeValue<"somedynamicenumerator">
    | TypedCustomAttributeValue<"somedate">
    | TypedCustomAttributeValue<"sometext">
  >;
  Milestone: Array<
    TypedCustomAttributeValue<"somedynamicenumerator">
  >;
}

export type TypedContextSubtype = keyof TypedContextSubtypeMap;

// Change TypedContext to have the new custom_attributes
export interface TypedContext<
  K extends TypedContextSubtype = TypedContextSubtype
> extends Omit<Context, "__entity_type__" | "__permissions" | "custom_attributes"> {
  __entity_type__: K;
  custom_attributes: TypedContextCustomAttributesMap[K]; // Not sure if this would be accepted, or we would need to change "K extends TypedContextSubtype = TypedContextSubtype" to something else. Not sure there is a way to tell TypeScript that the keys of two maps are the same.
  // [Rest of the properties from Context that are not omitted]
}
```

As I mentioned I'm not even sure it's possible and becomes a bit more involved, but if it sparks your imagination you can see if you can make something out of it.

ffMathy replied:
This has now been resolved.

ffMathy committed:
refactor: change types to not be based on omit

ffMathy committed:
Merge branch 'feature/use-custom-attribute-types-in-schema' of https://github.com/ffMathy/ftrack-ts-schema-generator into feature/use-custom-attribute-types-in-schema

ffMathy committed:
refactor: introduce TypeScript emitter and refactor code

ffMathy committed:
refactor: rename file

ffMathy committed:
refactor: better BasicLink fix

ffMathy committed:
fix: handle custom_attributes properly

ffMathy committed:
fix: formatting

ffMathy committed:
fix: tests

ffMathy commented:
@gismya I looked at your comments now.

Now, don't freak out when you see the diff. I really tried squeezing more stuff into the existing code, but I just felt like it was really hard for me to understand.

So to even be able to easily add these changes in the first place, and to make it easier to refactor and extend in the future, I've made some refactorings that look big at first glance, but are actually not.

I know this affects the diff quite a bit, but I feel confident about the result, and you can look at the highway test snapshot diff to see the total outcomes.

Ideally this would have been a separate PR first, but then this PR would have to be rewritten to merge into that etc.

Here's a summary:
- Made "__entity_type__" non-optional, so that type discriminators work correctly. I think we need a PR in the Ftrack API project too, to wrap all results in Optional<T> so we can further define which properties are actually present here.
- Introduced a TypeScriptEmitter. Its purpose is just a simple string builder with formatting capabilities. But it solves the problem of always returning strings and interpolating strings together. Now we can just pass the builder down the emit tree, and it has error capabilities too.
- Changed a few places from using granular functions to directly inlining them as multiline interpolated strings for more readability.
- Refactored the old Regex-based "BasicLink" hack and made it "stronger" (based on actually checking the schema's properties).
- Renamed "convertSchemaToInterface.ts" to "emitSchema.ts" to be consistent with the other files.

More refactoring can be done, but I wanted to focus on the minimal "breakthroughs" that could get me out of this situation I've been in for the past couple of PRs where it has been hard for me to make corrections into the existing code.

This isn't perfect, but it's the first steps.

ffMathy committed:
fix: custom_attributes should be an array type

ffMathy committed:
fix: bug where some types weren't mapped properly

ffMathy committed:
fix: make custom_attributes non-optional.

gismya commented on source/emitSchema.ts:
```diff
@@ -5,85 +5,112 @@ import type {
+  baseSchema: Schema | undefined,
+  schema: Schema
+) {
+  const omitList = ["__entity_type__", "__persmissions"];
```
Should probably not be `__persmissions` ;)

ffMathy replied:
Oooh 🤣 nice catch!

gismya commented on source/__snapshots__/default-highway-test.snap:
```diff
@@ -12,7 +11,7 @@ export interface Dashboard {
   name?: string;
   widgets?: DashboardWidget[];
-  __entity_type__?: "Dashboard";
+  __entity_type__: "Dashboard";
```
I'm not sure we can guarantee that the `__entity_type__` is always there. But maybe.

ffMathy replied:
In our tests, we have never encountered a scenario where it wasn't present. But that doesn't mean it doesn't exist.

gismya replied:
I haven't seen one either, but I've not gotten any commitments that we can always say it's there.
I'll try to get some answers on that.

ffMathy committed:
fix: typo

ffMathy committed:
fix: don't emit custom attributes and entity type as optionals

ffMathy commented:
Great comments - will reply soon to each of them :heart: Probably tomorrow. Didn't have much time today.

ffMathy commented:
> I'm worried about these two, because neither is required on write calls, and we are not entirely sure (but quite so) that it's always available on queries (And for custom_attributes it's only available if asked for).
Yeah you're right. I initially thought it wouldn't be bad because in the Ftrack API, if we wrap all of that in `Partial<T>`, it wouldn't matter much anyway.

Last night I used the TypeScript Playground for these tests. And there, it seemed to be acting up if the field was optional. However, now it seems even changing it to non-nullable doesn't make a difference.

> Could you give an example of what would not be working? Because type discriminators work with optional attributes in my tests.

Sure. Here's one of the cases I'd like to solve:

```typescript
const x: TypedContext = null!;
if(x.__entity_type__ === "Folder") {
  //i want TypeScript to understand that `x` is of type `Folder` here.
  //but it is not, so `custom_attributes` is also typed wrong.
}
```

I am working on another way to achieve the same, and trying out various things. In the meantime, do you have any guesses on how to solve it?

ffMathy commented:
Interesting. I found that if I do this trick, it works:

```typescript
type TypedContextSubtypeMapWithoutTypedContext = Omit<TypedContextSubtypeMap, 'TypedContext'>
type RealTypedContext = TypedContextSubtypeMapWithoutTypedContext[keyof TypedContextSubtypeMapWithoutTypedContext];

const x: RealTypedContext = null! as any;
if(x.__entity_type__ === "Folder") {
  //now inference works, and custom_attributes is typed correctly.
  x.custom_attributes.filter(x => x.key === "someboolean")
}
```

Will make a proposal based on this.

ffMathy committed:
fix: optional properties again

ffMathy committed:
fix: support type discriminators

ffMathy committed:
fix: ordering of BasicLink

ffMathy committed:
fix: ordering of EntityTypeMap

ffMathy committed:
fix: ordering of typed context types

ffMathy committed:
fix: ordering of custom attributes

ffMathy committed:
fix: more ordering tweaks for types

ffMathy committed:
fix: more ordering

gismya commented:
Wouldn't just removing `TypedContext` from `TypedContextSubtypeMap` have the same effect, without having to create multiple extra types on the way there?

ffMathy committed:
fix: ordering of priorities

ffMathy committed:
fix: statuses after priorities

ffMathy committed:
fix: don't export TypedContextForSubtype

ffMathy committed:
fix: final review fixes

ffMathy commented:
@gismya great point. That has been fixed now, and I also fixed the ordering. So now it should be good to go 🤞

ffMathy commented:
Wait, after switching away from the Omitted type again, type inference is once again broken.

ffMathy commented:
I think it's because a TypedContext implicitly sets `K` to `TypedContext`.

And we need TypedContext to be some kind of union. So for instance `TypedContextSubtypeMap[keyof TypedContextSubtypeMap]`, for it to know all the possibilities.

ffMathy committed:
Revert "fix: final review fixes"

This reverts commit 58d508a5a4c7d284a2f029aca2510fb591a77be6.

ffMathy committed:
fix: review fixes

ffMathy commented:
There we go, this should work. Can you give it another spin?

ffMathy committed:
fix: forgot to omit typed context

ffMathy commented:
@gismya I tried it now too with our codebase. Works flawlessly. Caught a few typing bugs too.

ffMathy commented:
Oh, and I forgot to comment on your GIF with the breathing bag 😆 That made me laugh out loud 😄 Glad you agree with the direction overall.

gismya reviewed (commented):
Trying to install and run in our environments, but lodash is installed as a commonjs module, and not ESM. Change to `lodash-es` or replace the functions.

ffMathy committed:
fix: switch to lodash-es

ffMathy commented:
@gismya done! Sorry for the inconvenience, and thanks for testing so fast.

gismya commented:
Does it make sense to use `void` for unused TypedContextCustomAttributeMaps?

![image](https://github.com/user-attachments/assets/8a3b04fa-3a10-4837-aa35-eb8b573903c3)

gismya commented on source/emitSchema.ts:
```diff
@@ -5,85 +5,109 @@ import type {
-      conversionErrors: [],
-    };
+    typescriptEmitter.appendCode(`
+      type ${interfaceName} = TypedContextForSubtype<"${interfaceName}">;
```
Not exporting this means you can't use them? Why was this done?

gismya replied:
When I manually add export to them, everything seems to work great again

ffMathy replied:
Yeah this was a pretty serious oversight - embarassing. I have fixed it now. Thank you.

ffMathy committed:
fix: missing export statement

ffMathy commented:
> Does it make sense to use `void` for unused TypedContextCustomAttributeMaps?
> 
> ![image](https://private-user-images.githubusercontent.com/8929526/358240748-8a3b04fa-3a10-4837-aa35-eb8b573903c3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM3MzU3OTksIm5iZiI6MTcyMzczNTQ5OSwicGF0aCI6Ii84OTI5NTI2LzM1ODI0MDc0OC04YTNiMDRmYS0zYTEwLTQ4MzctYWEzNS1lYjhiNTczOTAzYzMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgxNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MTVUMTUyNDU5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjAyMzJlNDNlYjZhNzIxYjM4OTRmZGY5NmYyZTRiOTIyY2IzYjhhNDdlMzJlMGRmMzVlY2M0ZjcwMDAwMGFjZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.vQRmbUpu4g6vYTufKF4WbyH0UCE7rrEYD55gmtRNxew)

I don't think so to be honest, no.

Actually, I don't think anyone will use the Map types. In my opinion they should actually not even be exported, but only used internally.

Can you elaborate on what you believe the value would be?

ffMathy committed:
fix: update snapshots

gismya commented:
> I don't think so to be honest, no.
> 
> Actually, I don't think anyone will use the Map types. In my opinion they should actually not even be exported, but only used internally.
> 
> Can you elaborate on what you believe the value would be?

They currently get exported as void if there are no custom attributes, thats what I'm saying. I was wondering if that was intentional 🙂

ffMathy committed:
fix: don't include types that don't have definitions

ffMathy committed:
fix: linting

ffMathy commented:
> > I don't think so to be honest, no.
> > Actually, I don't think anyone will use the Map types. In my opinion they should actually not even be exported, but only used internally.
> > Can you elaborate on what you believe the value would be?
> 
> They currently get exported as void if there are no custom attributes, thats what I'm saying. I was wondering if that was intentional 🙂

@gismya aaaah you're right. I was looking only at the highway snapshot and didn't notice that scenario. It has now been fixed, and you can see in the newest commits that it works as well (from looking at the snapshots of the other tests).

gismya commented:
Thinking another round there - wouldnt `never` be useful? Or would the end result be the same as not having anything there?

ffMathy commented:
Ah yeah maybe. Not sure. Maybe we can just pick one? As long as it works. Which one do you want?

gismya commented:
I'll test this way tomorrow, if it works fine it's fine by me 🙂

gismya commented:
It did not work, TypeScript throws errors if we reference things that don't exist in the map. `never` works fine though :)

ffMathy committed:
fix: never instead of not emitting

ffMathy commented:
You're right. I changed it to `never`. Fingers crossed now 🤞 

This will be a great change if it gets shipped. Looking forward to it.

gismya commented:
Fantastic job! I'll push this to a release immediately.

ffMathy commented:
WOOHOOOOO! And thanks! :heart:
